---
title: "Class 18: Mini Project - Investigating Pertussis Resurgence"
author: "Ellice Wang (PID: A16882742)"
date: March 9, 2025
toc: True
format: pdf
---

# 1. Investigating pertussis cases by year

> Q1. With the help of the R “addin” package datapasta assign the CDC pertussis case number data to a data frame called cdc and use ggplot to make a plot of cases numbers over time.

Installed "datapasta" using the command: install.packages("datapasta"). From the website "https://www.cdc.gov/pertussis/php/surveillance/pertussis-cases-by-year.html" I copy-pasted the data and navigated to: RStudio > Tools > Addins > Browse Addins > Paste as data.frame

```{r}
cdc <- data.frame(year = c(1922,1923,1924,1925,1926,1927,1928,1929,1930,
                    1931,1932,1933,1934,1935,1936,1937,1938,1939,1940,
                    1941,1942,1943,1944,1945,1946,1947,1948,1949,1950,
                    1951,1952,1953,1954,1955,1956,1957,1958,1959,1960,
                    1961,1962,1963,1964,1965,1966,1967,1968,1969,1970,
                    1971,1972,1973,1974,1975,1976,1977,1978,1979,1980,
                    1981,1982,1983,1984,1985,1986,1987,1988,1989,1990,
                    1991,1992,1993,1994,1995,1996,1997,1998,1999,2000,
                    2001,2002,2003,2004,2005,2006,2007,2008,2009,2010,
                    2011,2012,2013,2014,2015,2016,2017,2018,2019,2020,
                    2021,2022),
  no_reported_pertussis_cases = c(107473,164191,165418,152003,202210,181411,
                                   161799,197371,166914,172559,215343,179135,
                                   265269,180518,147237,214652,227319,103188,
                                   183866,222202,191383,191890,109873,
                                   133792,109860,156517,74715,69479,120718,
                                   68687,45030,37129,60886,62786,31732,28295,
                                   32148,40005,14809,11468,17749,17135,
                                   13005,6799,7717,9718,4810,3285,4249,
                                   3036,3287,1759,2402,1738,1010,2177,2063,
                                   1623,1730,1248,1895,2463,2276,3589,
                                   4195,2823,3450,4157,4570,2719,4083,6586,
                                   4617,5137,7796,6564,7405,7298,7867,
                                   7580,9771,11647,25827,25616,15632,10454,
                                   13278,16858,27550,18719,48277,28639,
                                   32971,20762,17972,18975,15609,18617,6124,
                                   2116,3044))
  
library(ggplot2)
num_cases_by_year <- ggplot(cdc) + 
  aes(year, no_reported_pertussis_cases) + 
  geom_point() + 
  geom_line() + 
  labs(title=("Pertussis Cases by Year (1922-2022)"), 
     x="Year", y="Number of Cases")
num_cases_by_year
```

# 2. A tale of two vaccines (wP & aP)

> Q2. Using the ggplot geom_vline() function add lines to your previous plot for the 1946 introduction of the wP vaccine and the 1996 switch to aP vaccine (see example in the hint below). What do you notice?

```{r}
num_cases_by_year + 
  geom_vline(xintercept=1946, linetype=2, color="blue") + 
  geom_vline(xintercept = 1996, linetype=2, color="red") + 
  annotate("text", x=1948.3, y=280000, label= "wP", color="blue") + 
  annotate("text", x=1998.3, y=280000, label= "aP", color="red") + 
  theme_bw()
```

> Q3. Describe what happened after the introduction of the aP vaccine? Do you have a possible explanation for the observed trend?

After the introduction of the aP vaccine, pertussis cases started increasing. I believe that part of the reason for this could be higher sensitivity to detecting whooping cough and increased sentiments discouraging vaccinations. 

# 3. Exploring CMI-PB data

```{r}
# Allows us to read, write and process JSON data
library(jsonlite)
subject <- read_json("https://www.cmi-pb.org/api/subject", 
                     simplifyVector = TRUE) 
head(subject, 3)
```

> Q4. How many aP and wP infancy vaccinated subjects are in the dataset?

There are 87 aP subjects and 85 wP subjects.

```{r}
table(subject$infancy_vac)
```

> Q5. How many Male and Female subjects/patients are in the dataset?

There are 112 female patients and 60 males. 

```{r}
table(subject$biological_sex)
```

> Q6. What is the breakdown of race and biological sex (e.g. number of Asian females, White males etc…)?

The breakdown is in the table below. 

```{r}
table(subject$biological_sex, subject$race)
```

## Side-Note: Working with dates

```{r}
library(lubridate)
today()
today() - ymd("2003-11-02")
time_length( today() - ymd("2003-11-02"),  "years")
```

> Q7. Using this approach determine (i) the average age of wP individuals, (ii) the average age of aP individuals; and (iii) are they significantly different?

i. average age of wP individuals: 36 years old
ii. average age of aP individuals: 27 years old
iii. signficantly different? They are 10 years apart; it is pretty different. 

```{r}
library(dplyr)

subject$age <- today() - ymd(subject$year_of_birth)
wp <- subject %>% filter(infancy_vac=="wP")
round(summary(time_length(wp$age, "years")))

ap <- subject %>% filter(infancy_vac=="aP")
round(summary(time_length(ap$age, "years")))
```

> Q8. Determine the age of all individuals at time of boost?

```{r}
diff_age <- ymd(subject$date_of_boost) - ymd(subject$year_of_birth)
age_at_boost <- time_length(diff_age, "year")
head(age_at_boost)
```

> Q9. With the help of a faceted boxplot or histogram (see below), do you think these two groups are significantly different?

These two groups are significantly different. aP subjects got a booster much quicker after initial vaccination than the wP group. 

```{r}
ggplot(subject) +
  aes(time_length(age, "year"),
      fill=as.factor(infancy_vac)) +
  geom_histogram(show.legend=FALSE) +
  facet_wrap(vars(infancy_vac), nrow=2) +
  xlab("Age in years")
```

## Joining multiple tables

```{r}
# Complete the API URLs...
specimen <- read_json("http://cmi-pb.org/api/v5_1/specimen", simplifyVector = TRUE) 
titer <- read_json("http://cmi-pb.org/api/v5_1/plasma_ab_titer", simplifyVector = TRUE) 

head(specimen)
```

> Q9. Complete the code to join specimen and subject tables to make a new merged data frame containing all specimen records along with their associated subject details:

```{r}
meta <- inner_join(specimen, subject)
dim(meta)
head(meta)
```

> Q10. Now using the same procedure join meta with titer data so we can further analyze this data in terms of time of visit aP/wP, male/female etc.

```{r}
abdata <- inner_join(titer, meta)
dim(abdata)
```

> Q11. How many specimens (i.e. entries in abdata) do we have for each isotype?

There are 6698 for IgE, 7265 for IgG, 11993 for IgG1, 12000 for IgG2, 12000 for IgG3, and 12000 for IgG4. 

```{r}
table(abdata$isotype)
```

> Q12. What are the different $dataset values in abdata and what do you notice about the number of rows for the most “recent” dataset?

I notice that there are more rows in the 2023 dataset than both the 2021 and 2022 but still less than 2020. 

```{r}
table(abdata$dataset)
```

# 4. Examine IgG Ab titer levels

```{r}
igg <- abdata %>% filter(isotype == "IgG")
head(igg)
```

> Q13. Complete the following code to make a summary boxplot of Ab titer levels (MFI) for all antigens:

```{r}
ggplot(igg) +
  aes(MFI_normalised, antigen) +
  geom_boxplot() + 
    xlim(0,75) +
  facet_wrap(vars(visit), nrow=2)
```

> Q14. What antigens show differences in the level of IgG antibody titers recognizing them over time? Why these and not others?

Most notably, FIM2/3 show differences in the level of IgG antibody titers recognizing them over time. This may be because there was increased specificity in the titration process. 

```{r}
ggplot(igg) +
  aes(MFI_normalised, antigen, col=infancy_vac ) +
  geom_boxplot(show.legend = FALSE) + 
  facet_wrap(vars(visit), nrow=2) +
  xlim(0,75) +
  theme_bw()

igg %>% filter(visit != 8) %>%
ggplot() +
  aes(MFI_normalised, antigen, col=infancy_vac ) +
  geom_boxplot(show.legend = FALSE) + 
  xlim(0,75) +
  facet_wrap(vars(infancy_vac, visit), nrow=2)
```

> Q15. Filter to pull out only two specific antigens for analysis and create a boxplot for each. You can chose any you like. Below I picked a “control” antigen (“OVA”, that is not in our vaccines) and a clear antigen of interest (“PT”, Pertussis Toxin, one of the key virulence factors produced by the bacterium B. pertussis).

```{r}
head(igg)
```


```{r}
filter(igg, antigen=="PT") %>%
  ggplot() +
  aes(MFI_normalised, col=infancy_vac) +
  geom_boxplot(show.legend = T) +
  facet_wrap(vars(visit)) +
  theme_bw() + 
  labs(title="PT Antigen Levels per Visit")
```

```{r}
filter(igg, antigen=="FIM2/3") %>%
  ggplot() +
  aes(MFI_normalised, col=infancy_vac) +
  geom_boxplot(show.legend = T) +
  facet_wrap(vars(visit)) +
  theme_bw() + 
  labs(title="FIM 2/3 Antigen Levels per Visit")
```

> Q16. What do you notice about these two antigens time courses and the PT data in particular?

Both PT and FIM2/3 levels rise over time. PT data seems to peak around visit 8 while FIM 2/3 data peaks around day 5. 

> Q17. Do you see any clear difference in aP vs. wP responses?

Not any clear differences though it seems as if aP has a broader range for the FIM2/3 antigen while wP has a broader range for the PT antigen. 

```{r}
abdata.21 <- abdata %>% filter(dataset == "2021_dataset")

abdata.21 %>% 
  filter(isotype == "IgG",  antigen == "PT") %>%
  ggplot() +
    aes(x=planned_day_relative_to_boost,
        y=MFI_normalised,
        col=infancy_vac,
        group=subject_id) +
    geom_point() +
    geom_line() +
    geom_vline(xintercept=0, linetype="dashed") +
    geom_vline(xintercept=14, linetype="dashed") +
  labs(title="2021 dataset IgG PT",
       subtitle = "Dashed lines indicate day 0 (pre-boost) and 14 (apparent peak levels)")
```

> Q18. Does this trend look similar for the 2020 dataset?

```{r}
abdata.20 <- abdata %>% filter(dataset == "2020_dataset")

abdata.20 %>% 
  filter(isotype == "IgG",  antigen == "PT") %>%
  ggplot() +
    aes(x=planned_day_relative_to_boost,
        y=MFI_normalised,
        col=infancy_vac,
        group=subject_id) +
    geom_point() +
    geom_line() +
    geom_vline(xintercept=0, linetype="dashed") +
    geom_vline(xintercept=14, linetype="dashed") +
  labs(title="2020 dataset IgG PT",
       subtitle = "Dashed lines indicate day 0 (pre-boost) and 14 (apparent peak levels)") + 
  xlim(0,125)
```

There is not a clear divide between aP and wP as seen in the 2021 dataset, in which the wP data clearly had a higher MFI normalized score. The 2020 dataset looks a bit messier overall. 
# 5. Obtaining CMI-PB RNASeq data

```{r}
url <- "https://www.cmi-pb.org/api/v2/rnaseq?versioned_ensembl_gene_id=eq.ENSG00000211896.7"

rna <- read_json(url, simplifyVector = TRUE) 

ssrna <- inner_join(rna, meta)
```

> Q19. Make a plot of the time course of gene expression for IGHG1 gene (i.e. a plot of visit vs. tpm).

```{r}
ggplot(ssrna) +
  aes(visit, tpm, group=subject_id) +
  geom_point() +
  geom_line(alpha=0.2)
```

> Q20.: What do you notice about the expression of this gene (i.e. when is it at it’s maximum level)?

Maximum expression level for IGH1 generally occurs at visit 4. It increases in the first couple of visits until visit 4, then generally decreases for the next visits 

> Q21. Does this pattern in time match the trend of antibody titer data? If not, why not?

It is pretty consistent with the trend of antibody titer data, in regards to where antigens levels peak around visit 4/5. However the difference with the antibody titer data is that the levels do not drop off after the visits 4/5. This is because after the antibody gene has been expressed, the antibody is able to stay in the cell for a longer period of time. 

```{r}
ggplot(ssrna) +
  aes(tpm, col=infancy_vac) +
  geom_boxplot() +
  facet_wrap(vars(visit))

ssrna %>%  
  filter(visit==4) %>% 
  ggplot() +
    aes(tpm, col=infancy_vac) + geom_density() + 
    geom_rug() 
```

